src/proc/validator.rs

```diff
@@ -690,7 +695,7 @@ impl Validator {
     }
 
     /// Check the given module to be valid.
-    pub fn validate(&mut self, module: &crate::Module) -> Result<(), ValidationError> {
+    pub fn validate(&mut self, module: &crate::Module) -> Result<Analysis, ValidationError> {
```

Comment:
We are currently doing a ton of extra work, already, because we've been trying to totally conceal the byproducts of the IR from the user. `Typifier` is the extreme case - it's used in all the things. First, in front-ends, then in the validator, and finally in the backends. Technically, there should only be need to one of them. Maximum - two, if the IR undergoes any processing. But we have 3 even without any processing.

So I think exposing the `Analysis` in the API makes sense. It's a byproduct that we *can* replicate if want, but we shouldn't need to if the user can give it back to us.

Reply:
Yep, agreed we should reduce the extra work by storing the intermediate results when possible. I was thinking about wrapping the intermediate state into a struct somehow, to allow us to run the passes internally and save any intermediate results that are useful to later passes. So in that case we'd be able to pass `Typifier` all the way from the frontend to the backends internally without the user being aware of `Typifier`

Reply:
I think `Analysis` can be this struct. It already combines information about different things. We could do the following:
  - make `Analysis::new` to accept a typifier struct
  - copy the type information from it into `ExpressionInfo`
  - make the backends use the analysis info instead of re-generating the typifier

Reply:
:+1: sounds good to me!

Reply:
Moving the follow-up to https://github.com/gfx-rs/naga/issues/260#issuecomment-778824692